File: LatihanInheritance/kendaraan.cpp
#include "kendaraan.hpp"

Kendaraan::Kendaraan()
{
    this->nomor = 0;
    this->tahun = 0;
    this->merk = "XXX";
    this->kategori = "mobil";
}

Kendaraan::Kendaraan(int nomor, string kategori, string merk, int tahun)
{
    this->nomor = nomor;
    this->tahun = tahun;
    this->merk = merk;
    this->kategori = kategori;
}

Kendaraan::Kendaraan(const Kendaraan &k)
{
    this->nomor = k.nomor;
    this->tahun = k.tahun;
    this->merk = k.merk;
    this->kategori = k.kategori;
}

Kendaraan::~Kendaraan()
{
    // Destructor
}

void Kendaraan::printInfo()
{
    cout << "Nomor: " << this->nomor << endl;
    cout << "Merk: " << this->merk << endl;
    cout << "Kategori: " << this->kategori << endl;
    cout << "Tahun: " << this->tahun << endl;
}

int Kendaraan::biayaSewa(int lamaSewa)
{
    int ans = 0;
    if (this->kategori == "Bus")
    {
        ans = 1000000 * lamaSewa;
    }
    else if (this->kategori == "Minibus")
    {
        if (lamaSewa <= 5)
        {
            ans = 5000000;
        }
        else
        {
            ans = 5000000 + 500000 * (lamaSewa - 5);
        }
    }
    else if (this->kategori == "Mobil")
    {
        ans = 500000 * lamaSewa;
    }
    return ans;
}

KoleksiKendaraan::KoleksiKendaraan() : size(100)
{
    this->Neff = 0;
    this->buffer = new Kendaraan[this->size];
}

KoleksiKendaraan::KoleksiKendaraan(int s) : size(s)
{
    this->Neff = 0;
    this->buffer = new Kendaraan[this->size];
}

KoleksiKendaraan::KoleksiKendaraan(const KoleksiKendaraan &k) : size(k.size)
{
    this->Neff = k.Neff;
    this->buffer = new Kendaraan[this->size];
    for(int i = 0; i < this->Neff; i++){
        this->buffer[i] = k.buffer[i];
    }
}

KoleksiKendaraan::~KoleksiKendaraan()
{
    // Destructor
}

void KoleksiKendaraan::printAll(){
    for(int i = 0; i < this->Neff; i++){
        cout << "Kendaraan ke-" << i+1 << ":" << endl;
        this->buffer[i].printInfo();
    }
}

void KoleksiKendaraan::operator<<(Kendaraan k){
    this->buffer[Neff] = k;
    this->Neff++;
}

void KoleksiKendaraan::operator<<(KoleksiKendaraan k){
    for(int i = 0; i < k.Neff; i++){
        this->buffer[Neff] = k.buffer[i];
        this->Neff++;
    }
}

File: Latihan/Kendaraan.java
package Latihan;

abstract class Kendaraan {
  protected int nomor;
  protected int tahun;
  protected String merk;

  public Kendaraan() {
    this.nomor = 0;
    this.tahun = 0;
    this.merk = "";
  }

  public Kendaraan(int nomor, int tahun, String merk) {
    this.nomor = nomor;
    this.tahun = tahun;
    this.merk = merk;
  }

  public Kendaraan(Kendaraan k) {
    this.nomor = k.nomor;
    this.tahun = k.tahun;
    this.merk = k.merk;
  }

  public void printInfo() {
    System.out.println("Nomor Kendaraan: " + nomor);
    System.out.println("Tahun Kendaraan: " + tahun);
    System.out.println("Merk Kendaraan: " + merk);
  }

  abstract int biayaSewa(int lamaSewa);
}

public class Bus extends Kendaraan {
  private int kapasitas;

  public Bus() {
    super();
    this.kapasitas = 0;
  }

  public Bus(int nomor, int tahun, String merk, int kapasitas) {
    super(nomor, tahun, merk);
    this.kapasitas = kapasitas;
  }

  public Bus(Bus b) {
    super(b);
    this.kapasitas = b.kapasitas;
  }

  public void printInfo() {
    super.printInfo();
    System.out.println("Kapasitas Bus: " + kapasitas);
    System.out.println("Kategori: Bus");
  }

  public int biayaSewa(int lamaSewa) {
    return lamaSewa * 10000;
  }
}

public class MiniBus extends Kendaraan {
  // private int kapasitas;

  public MiniBus() {
    super();
  }

  public MiniBus(int nomor, int tahun, String merk, int kapasitas) {
    super(nomor, tahun, merk);
  }

  public MiniBus(MiniBus m) {
    super(m);
  }

  public void printInfo() {
    super.printInfo();
    System.out.println("Kategori: Mini Bus");
  }

  public int biayaSewa(int lamaSewa) {
    if (lamaSewa <= 5) {
      return 5000000;
    } else {
      return 5000000 + 500000 * (lamaSewa - 5);
    }
  }
}

public class Mobil extends Kendaraan {
  private String supir;

  public Mobil() {
    super();
    this.supir = "";
  }

  public Mobil(int nomor, int tahun, String merk, String supir) {
    super(nomor, tahun, merk);
    this.supir = supir;
  }

  public Mobil(Mobil m) {
    super(m);
    this.supir = m.supir;
  }

  public void printInfo() {
    super.printInfo();
    System.out.println("Nama supir: " + this.supir);
    System.out.println("Kategori: Mobil");
  }

  public int biayaSewa(int lamaSewa) {
    return lamaSewa * 500000;
  }
}

public class KoleksiKendaraan {
  private int size;
  private int neff;
  private Kendaraan buffer[];

  public KoleksiKendaraan() {
    this.size = 10;
    this.neff = 0;
    this.buffer = new Kendaraan[this.size];
  }

  public KoleksiKendaraan(int size) {
    this.size = size;
    this.neff = 0;
    this.buffer = new Kendaraan[this.size];
  }

  public KoleksiKendaraan(KoleksiKendaraan kk){
    this.size = kk.size;
    this.neff = kk.neff;
    this.buffer = new Kendaraan[this.size];
    for (int i = 0; i < this.neff; i++) {
      this.buffer[i] = kk.buffer[i];
    }
  }

  public void printAll(){
    for (int i = 0; i < this.neff; i++) {
      this.buffer[i].printInfo();
    }
  }

  public void addKendaraan(Kendaraan k){
    if (this.neff == this.size) {
      Kendaraan temp[] = new Kendaraan[this.size * 2];
      for (int i = 0; i < this.neff; i++) {
        temp[i] = this.buffer[i];
      }
      this.buffer = temp;
      this.size *= 2;
    }
    this.buffer[this.neff] = k;
    this.neff++;
  }

  public void addKoleksiKendaraan(KoleksiKendaraan kk){
    if (this.neff + kk.neff > this.size) {
      Kendaraan temp[] = new Kendaraan[this.size * 2];
      for (int i = 0; i < this.neff; i++) {
        temp[i] = this.buffer[i];
      }
      this.buffer = temp;
      this.size *= 2;
    }
    for (int i = 0; i < kk.neff; i++) {
      this.buffer[this.neff] = kk.buffer[i];
      this.neff++;
    }
  }
}